test/extensions/filters/http/common/fuzz/filter_fuzz_test.cc

```diff
@@ -0,0 +1,106 @@
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+
+class UberFilterFuzzer {
+public:
+  UberFilterFuzzer() {
+    ON_CALL(filter_callback_, addStreamDecoderFilter(_))
+        .WillByDefault(
+            Invoke([&](std::shared_ptr<Envoy::Http::StreamDecoderFilter> filter) -> void {
+              filter_ = filter;
+              filter_->setDecoderFilterCallbacks(callbacks_);
+            }));
+  }
+
+  // This executes the methods to be fuzzed.
+  void decode(Http::StreamDecoderFilter* filter, const test::fuzz::HttpData& data) {
+    ENVOY_LOG_MISC(info, "Decoding {} with filter", data.DebugString());
+    bool end_stream = false;
+
+    Http::TestHeaderMapImpl headers = Fuzz::fromHeaders(data.headers());
+    if (data.data().size() == 0 && !data.has_trailers()) {
+      end_stream = true;
+    }
+    filter->decodeHeaders(headers, end_stream);
+
+    for (int i = 0; i < data.data().size(); i++) {
+      if (i == data.data().size() - 1 && !data.has_trailers()) {
+        end_stream = true;
+      }
+      Buffer::OwnedImpl buffer(data.data().Get(i));
+      filter->decodeData(buffer, end_stream);
+    }
+
+    if (data.has_trailers()) {
+      Http::TestHeaderMapImpl trailers = Fuzz::fromHeaders(data.trailers());
+      filter->decodeTrailers(trailers);
+    }
+  }
+
+  // This creates the filter from the proto and runs decode.
+  void
+  fuzz(const envoy::config::filter::network::http_connection_manager::v2::HttpFilter& proto_config,
+       const test::fuzz::HttpData& data) {
+    // TODO: Dissociate filter creation from active stream.
+    auto& factory =
+        Config::Utility::getAndCheckFactory<Server::Configuration::NamedHttpFilterConfigFactory>(
+            proto_config.name());
+    ProtobufTypes::MessagePtr message = Config::Utility::translateToFactoryConfig(
+        proto_config, Envoy::ProtobufMessage::getNullValidationVisitor(), factory);
+    Http::FilterFactoryCb cb =
+        factory.createFilterFactoryFromProto(*message, "fuzz", factory_context_);
+
+    // Set the filter_ and execute the code to  be fuzzed.
+    cb(filter_callback_);
+    decode(filter_.get(), data);
+  }
+
+  NiceMock<Server::Configuration::MockFactoryContext> factory_context_;
+  NiceMock<Http::MockStreamDecoderFilterCallbacks> callbacks_;
+  NiceMock<Http::MockFilterChainFactoryCallbacks> filter_callback_;
+  std::shared_ptr<Http::StreamDecoderFilter> filter_;
+};
+
+DEFINE_PROTO_FUZZER(const test::extensions::filters::http::FilterFuzzTestCase& input) {
+  // TODO: Hard-coded right now. Replace with pulling from a directory.
+  const std::string config_string =
+      R"EOF(
+name: envoy.buffer
+typed_config:
+    "@type": type.googleapis.com/envoy.config.filter.http.buffer.v2.Buffer
+    max_request_bytes : 5242880
+)EOF";
+  // Create proto_config from the string.
+  envoy::config::filter::network::http_connection_manager::v2::HttpFilter proto_config;
+  TestUtility::loadFromYaml(config_string, proto_config);
```

Comment:
I like the new approach to making this generic. What I think we can do to make this more robust is:
1. Use some initial entropy to index into the list of available HTTP filters. There's probably no convenient integer index mechanism right now, but it should be easy to add to the registry.
2. Once we have the filter factory, we can then ask it it `createEmptyConfigProto()`. This will give us a `Protobuf::Message`.
3. With the message obtained in (2) (and/or its descriptor), we ask `libprotobuf-mutator` to populate with structurally valid content.
4. We fuzz with `UberFilterFuzzer` as in this PR, that bit looks great.

Reply:
All set. Here are some things:
3. The mutator sometimes produces an invalid config, and so this code is in a try/catch block. I'm wondering if it's better to run this a few times rather than abort the fuzz test immediately, although I think it doesn't make a difference.
4. The class is starting to get complicated. I could move it out in to a library.